pkg/awsclient/elb_helper.go

```diff
@@ -0,0 +1,221 @@
+// the kube-api
+func (c *awsClient) addListenersToELB(elbName string, listeners []*elb.Listener) error {
+	i := &elb.CreateLoadBalancerListenersInput{
+		Listeners:        listeners,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.CreateLoadBalancerListeners(i)
+	return err
+}
+
+// AddLoadBalancerInstances will attach +instanceIds+ to +elbName+
+// so that they begin to receive traffic. Note that this takes an amount of
+// time to return. This is also additive (but idempotent - TODO: Validate this).
+// Note that the recommended steps:
+// 1. stop the instance,
+// 2. deregister the instance,
+// 3. start the instance,
+// 4. and then register the instance.
+func (c *awsClient) AddLoadBalancerInstances(elbName string, instanceIds []string) error {
+	instances := make([]*elb.Instance, 0)
+	for _, instance := range instanceIds {
+		instances = append(instances, &elb.Instance{InstanceId: aws.String(instance)})
+	}
+	i := &elb.RegisterInstancesWithLoadBalancerInput{
+		Instances:        instances,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.RegisterInstancesWithLoadBalancer(i)
+	return err
+}
+
+// RemoveInstancesFromLoadBalancer removes +instanceIds+ from +elbName+, eg when an Node is deleted.
+func (c *awsClient) RemoveInstancesFromLoadBalancer(elbName string, instanceIds []string) error {
+	instances := make([]*elb.Instance, 0)
+	for _, instance := range instanceIds {
+		instances = append(instances, &elb.Instance{InstanceId: aws.String(instance)})
+	}
+	i := &elb.DeregisterInstancesFromLoadBalancerInput{
+		Instances:        instances,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.DeregisterInstancesFromLoadBalancer(i)
+	return err
+}
+
+// DoesELBExist checks for the existence of an ELB by name. If there's an AWS
+// error it is returned.
+func (c *awsClient) DoesELBExist(elbName string) (bool, *AWSLoadBalancer, error) {
+	i := &elb.DescribeLoadBalancersInput{
+		LoadBalancerNames: []*string{aws.String(elbName)},
+	}
+	res, err := c.DescribeLoadBalancers(i)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case elb.ErrCodeAccessPointNotFoundException:
+				return false, &AWSLoadBalancer{}, nil
+			default:
+				return false, &AWSLoadBalancer{}, err
+			}
+		}
+	}
+	return true, &AWSLoadBalancer{ELBName: elbName, DNSName: *res.LoadBalancerDescriptions[0].DNSName, DNSZoneId: *res.LoadBalancerDescriptions[0].CanonicalHostedZoneNameID}, nil
+}
+
+// LoadBalancerInfo fields needed from NLBs
+type LoadBalancerInfo struct {
+	ExtCanonicalHostedZoneNameID string
+	ExtDNSName                   string
+	ExtLoadBalancerArn           string
+	ExtLoadBalancerName          string
+	ExtScheme                    string
+	IntCanonicalHostedZoneNameID string
+	IntDNSName                   string
+	IntLoadBalancerArn           string
+	IntLoadBalancerName          string
+	IntScheme                    string
+}
```

Comment:
How about treating this like a list so that in lines 182-194 it's dealing with a list of all the external load balancers, which is returned to the caller. It might be that we need more information about the load balancer, such as the ports, or whatever else to be sure that we eventually delete the correct one, in case there's more than the one external nlb. (I also changed the name just for clarity in the snippet)

```go
// LoadBalancerV2 is a list of non-classic AWS ELBs
type LoadBalancerV2 struct {
  HostedZoneNameId string
  DNSName          string
  Arn              string
  Name             string
  Scheme           string
// ...
}
```

With this kind of structure, there's less duplication in the structure.

Later, the loop could look like:

```go
loadBalancers := make([]LoadBalancersV2, 0)
for _, loadBalancer := range output.LoadBalancers {
	loadBalancers = append(loadBalancers, LoadBalancersV2{
		HostedZoneNameId: loadBalancer.CanonicalHostedZoneId,
		DNSName:          loadBalancer.DNSName,
		Arn:              loadBalancer.LoadBalancerArn,
		Name:             loadBalancer.LoadBalancerName,
		Scheme:           loadBalancer.Scheme,
	})
}

return loadBalancers, nil
```

Since this method is currently mainly used to delete the external NLB, that caller could range through the list to find the appropriate one and perform the deletion before `break`ing out of the loop.

Reply:
That feels very AWS specific to me. Since we don't know what other cloud api looks like does it make more sense to refactor later on?

Reply:
I think it makes sense to handle it now as a code clarity change. Remember, this is the aws client helper, so it makes sense to be aws-specific here. Once we get to the multi-cloud world we already have to refactor AWS-specific parts out, so we can take that time do it. Here, in our AWS library, we can be as AWS-specific as we like.

Reply:
we use this method to both delete the external NLB but also get some information about the internal NLB for `upsertCNAME` func. if we use this method and I imagine in our controller we would have to put AWS specific logic like `scheme == internal`. I was trying to avoid having to do this

Reply:
That's true. What if the controller is calling something like `DeleteExternalLoadBalancer` and passes in what the name of that external load balancer is? It could come from the `Machine` object. In this way the controller doesn't need to know what kind of load balancers are under the hood?